### ISSUE TITLE
multithreaded ncvisual_stream()

### ISSUE THREAD

**Issue by dankamongmen** [dankamongmen/notcurses#1620]
Similarly to work in `xray` and planned work in `yield`, `ncplayer` ought use two threads for multiframe rendering. One will rasterize (write), and one will decode the next frame. This ought be a nice pickup for pixel mode, and allow more precision in all modes. It will be important to retain timing information here, which we're not yet doing in `xray`. Maybe it would be best to embed threading into `ncplayer_stream()/ncdirect_stream()`?

**Comment by dankamongmen** [comment]
This *requires* embedding the threading into `ncplayer_stream()`. No two ways around it. And that means we'd need one per media implementation, though technically only ffmpeg could benefit from this right now...it ought be doable, though.

I don't think there's much point in exposing this as an option. We'll just always do it. Even on a single-core machine, we'd be combining computation (decode) with blocking I/O (raster write), so it ought benefit.

**Comment by dankamongmen** [comment]
So one thread can definitely decode a frame while the other is doing arbitrary stuff. In `xray`, we also allow the thread to render to an (off-pile, new) frame, to maximize the parallelism. This is fine for the case where we're not rendering to a specified plane, but if we're using a specified plane, we'd have to hold off on the rendering. That still ought be a nice little speed advantage. The issue is that if rendering isn't done inside the locked sequence, we can't safely decode a new frame (it overwrites the previous frame).

**Comment by dankamongmen** [comment]
when this is done, i think we ought be able to remove all threading from `xray`, and just rely on this, no? which would render #1686 immaterial, or at least mean it need be fixed in `ncvisual_stream()`.

**Comment by dankamongmen** [comment]
so yeah, it seems all we can safely do is parallelize the decoding against the rendering.

**Comment by dankamongmen** [comment]
hrmmm, i don't think we can even do that. we pass the `ncvisual` down to the streamer callback, which can do what it wants to it. if we ran the next decode in parallel, we could break any references being made to it in said callback (e.g. subtitle extraction).

well, shit.

we could of course create a thread and duplicate the input source, and have each decode two frames for each one they display, but that's a good bit of extra work. we could get around this double decode by having one thread decoding and writing to an off-pile plane, and another thread just reparenting that plane and bringing it to the current pile. we should probably do that in `[xray]`. we'd still need to avoid touching the `ncvisual` in the callback.

maybe a flag to indicate "i'm not going to touch the `ncvisual`"? eh. we'd always want to be able to get the subtitle plane. maybe that ought just be part of `ncvisual_stream()`, and a new parameter to the streamer callback.

so, we're not getting this done in 3.0. maybe we can make that change to `[xray]`, though, and eliminate the double work. hmmm technically the decoder thread could whip through decoding as many frames as it wanted, and just hand planes to the rasterizer thread. mmmmm, yes!

ok, i like that idea. and we could use it in `ncvisual_stream()` if we just changed the callback API to not take the `ncvisual`, but instead (two) `ncplane`s. except if they specified the plane...maybe we ought require an `ncplane_options` be provided, and refuse a specified `n`? or maybe only do a threaded solution with one?

how would this work with TAM reapplication? erp, i don't think that it would? =[

maybe we just don't do this in `ncvisual_stream()`. if you want a threaded solution, you do one. or we add a function explicitly for it. yeah =\.

in fact, how do we even get TAM reapplication in the current `[xray]` scheme? i guess we don't? hrmmm.

alright, closing this one. sad!
